File: src/shared/dataServer/ndjson.ts
// NDJSON framing — one JSON message per LF-terminated line.
//
// THIS IS THE ONLY FILE IN src/shared/dataServer THAT KNOWS A NEWLINE EXISTS, and keeping it that
// way is the whole point of the transport seam (see transport.ts). If the wire becomes WebSocket
// frames over the open internet, this file gains a sibling and nothing else in the tree changes.
// `tests/dataServerTransport.test.mts` asserts the absence directly, by reading every other file
// in this directory and failing on a newline literal.
//
// WHY LF IS SAFE AS A DELIMITER, written down so nobody re-derives it nervously: `JSON.stringify`
// escapes every control character inside a string, so a serialized message can never contain a raw
// newline however hostile its contents — a `\n` in a row's text arrives on the wire as the two
// characters `\` and `n`. That is a property of JSON, not of this app's data, so no amount of
// game-log weirdness can smuggle a frame boundary through.
//
// THE DELIMITER IS LF, NEVER CRLF. Windows is the only platform this app ships on and it is
// exactly where a text-mode stream would helpfully translate one into the other; a trailing `\r`
// is stripped on decode rather than trusted, so a peer that framed with CRLF is still read
// correctly.

import { TransportError, type Transport } from './transport'

/** The frame delimiter. One character, and the only one this protocol treats as structural. */
export const DELIMITER = '\n'

/**
 * The largest single frame this codec will assemble, in characters.
 *
 * A FRAMING guard, not a protocol rule: a peer that never sends a delimiter would otherwise grow
 * the buffer without bound, which is a denial of service a loopback socket makes trivial. Payload
 * budgets — how big a view's window may be — are a protocol concern and live engine-side, nowhere
 * near this number. It matches `MAX_LINE_BYTES` in the Rust codec.
 */
export const MAX_LINE_CHARS = 8 * 1024 * 1024

/** Serialize one message into its wire form: the JSON, then the delimiter. */
export function encodeLine(message: unknown): string {
  let json: string
  try {
    json = JSON.stringify(message)
  } catch (cause) {
    throw new TransportError('encode', 'message could not be serialized', cause)
  }
  if (json === undefined) throw new TransportError('encode', 'message serialized to nothing')
  return `${json}${DELIMITER}`
}

/**
 * Parse one wire line back into a message. The line must NOT carry its delimiter.
 *
 * It returns `unknown` rather than a caller-chosen type on purpose: a JSON parse cannot know what
 * arrived, and a generic here would be a cast wearing a type signature. The narrowing belongs to
 * whoever knows which direction the connection runs — see `createNdjsonTransport`.
 */
export function decodeLine(line: string): unknown {
  const trimmed = line.endsWith('\r') ? line.slice(0, -1) : line
  try {
    return JSON.parse(trimmed)
  } catch (cause) {
    throw new TransportError('decode', 'a frame was not a JSON message', cause)
  }
}

/**
 * The stateful half: bytes arrive in whatever chunks the OS felt like, and a message may be split
 * across any number of them. `push` returns the COMPLETE lines it now has, keeping the remainder.
 *
 * It is separate from the transport so it can be tested on its own against adversarial chunking —
 * a codec that only works when each read happens to land on a frame boundary is a codec that works
 * on a test double and fails on a socket.
 */
export class LineDecoder {
  private buffer = ''

  constructor(private readonly maxLineChars = MAX_LINE_CHARS) {}

  /** Feed a chunk; get back every whole line it completed, in order. */
  push(chunk: string): string[] {
    this.buffer += chunk
    const lines: string[] = []
    let at = this.buffer.indexOf(DELIMITER)
    while (at !== -1) {
      lines.push(this.buffer.slice(0, at))
      this.buffer = this.buffer.slice(at + 1)
      at = this.buffer.indexOf(DELIMITER)
    }
    if (this.buffer.length > this.maxLineChars) {
      this.buffer = ''
      throw new TransportError('frameTooLarge', `a frame exceeded ${String(this.maxLineChars)} characters`)
    }
    return lines
  }

  /**
   * What is still buffered. A non-empty remainder at end of stream is a TRUNCATED FRAME — half a
   * message discarded in silence is how a client ends up rendering a world nobody sent — so the
   * transport reports it rather than dropping it.
   */
  get pending(): string {
    return this.buffer
  }
}

/** What an NDJSON transport needs from whatever is actually carrying the bytes. */
export interface ByteChannel {
  /** Push bytes at the peer. */
  write(chunk: string): void
  /** Register the one handler for arriving bytes. */
  onData(handler: (chunk: string) => void): void
  /** Register the handler for the stream ending, cleanly or otherwise. */
  onClose(handler: (error?: unknown) => void): void
  /** Stop. Idempotent. */
  close(): void
}

/**
 * A {@link Transport} over any byte channel.
 *
 * It takes a channel rather than a socket because phase 0 has no socket: the suite drives it over
 * an in-memory pipe, and the supervisor will hand it a `net.Socket` adapter without this file
 * changing.
 */
export function createNdjsonTransport<Out, In>(channel: ByteChannel): Transport<Out, In> {
  const decoder = new LineDecoder()
  let onMessage: ((message: In) => void) | undefined
  let onError: ((error: TransportError) => void) | undefined
  let closed = false

  const fail = (error: TransportError): void => {
    if (closed) return
    closed = true
    channel.close()
    onError?.(error)
  }

  channel.onData((chunk) => {
    if (closed) return
    let lines: string[]
    try {
      lines = decoder.push(chunk)
    } catch (e) {
      fail(e instanceof TransportError ? e : new TransportError('io', 'framing failed', e))
      return
    }
    for (const line of lines) {
      if (closed) return
      // A blank line is not a message. Being lenient here costs nothing and makes a peer that
      // ends a batch with an extra newline readable instead of fatal.
      if (line.trim() === '') continue
      try {
        // THE ONE CAST IN THE SEAM, and it is where it belongs: the transport is the only thing
        // that knows which direction this connection runs, so it is the only thing entitled to say
        // what a decoded frame is. Whether the peer actually honoured the contract is the schema's
        // question, not the codec's.
        onMessage?.(decodeLine(line) as In)
      } catch (e) {
        fail(e instanceof TransportError ? e : new TransportError('decode', 'a frame was refused', e))
        return
      }
    }
  })

  channel.onClose((error) => {
    if (closed) return
    if (decoder.pending.trim() !== '') {
      fail(new TransportError('decode', 'the stream ended mid-frame'))
      return
    }
    closed = true
    if (error !== undefined) onError?.(new TransportError('io', 'the transport failed', error))
  })

  return {
    send(message) {
      if (closed) throw new TransportError('closed', 'the transport is closed')
      channel.write(encodeLine(message))
    },
    onMessage(handler) {
      onMessage = handler
    },
    onError(handler) {
      onError = handler
    },
    close() {
      if (closed) return
      closed = true
      channel.close()
    },
    get closed() {
      return closed
    }
  }
}

File: engine/crates/protocol/src/transport/ndjson.rs
//! NDJSON framing — one JSON message per LF-terminated line. The only module in the crate that
//! knows a newline exists; a different wire is a sibling of this file.
//!
//! LF is safe as a delimiter because `serde_json` escapes every control character inside a string,
//! so a serialized message can never contain a raw newline however hostile its contents. That is a
//! property of JSON, not of this app's data.
//!
//! The delimiter is LF, never CRLF: every stream this transport touches is opened in binary, and a
//! trailing `\r` is stripped on decode rather than trusted, so a CRLF-framing peer still reads
//! correctly.

use std::io::{BufRead, Write};
use std::marker::PhantomData;

use serde::de::DeserializeOwned;
use serde::Serialize;

use super::{Transport, TransportError};

/// The frame delimiter. One byte, and the only one this crate treats as structural.
pub const DELIMITER: u8 = b'\n';

/// The largest single frame this transport will assemble, in bytes.
///
/// A framing guard, not a protocol rule: a peer that never sends a delimiter would otherwise grow
/// the read buffer without bound. Payload budgets are a protocol concern and live engine-side.
/// 8 MiB is far above any legitimate message and far below anything that threatens the process.
pub const MAX_LINE_BYTES: usize = 8 * 1024 * 1024;

/// Serialize one message into its wire form: the JSON, then the delimiter.
///
/// # Errors
/// [`TransportError::Encode`] if the message will not serialize.
pub fn encode_line<T: Serialize>(message: &T) -> Result<String, TransportError> {
    let mut line = serde_json::to_string(message).map_err(TransportError::Encode)?;
    line.push(char::from(DELIMITER));
    Ok(line)
}

/// Parse one wire line back into a message. The line must not carry its delimiter; a trailing `\r`
/// is tolerated so a CRLF-framing peer is still understood.
///
/// # Errors
/// [`TransportError::Decode`] if the line is not a message of this type.
pub fn decode_line<T: DeserializeOwned>(line: &str) -> Result<T, TransportError> {
    let trimmed = line.strip_suffix('\r').unwrap_or(line);
    serde_json::from_str(trimmed).map_err(TransportError::Decode)
}

/// Parse one complete frame's bytes back into a message — what [`NdjsonTransport::recv`] uses.
///
/// The text conversion happens exactly once, over a whole frame, for correctness rather than speed:
/// a UTF-8 character is up to four bytes and a read boundary falls wherever the OS puts it, so a
/// character routinely straddles two reads. Decoding each read separately (`from_utf8_lossy`)
/// replaces each half with U+FFFD, and because both halves sit inside a JSON string the corrupted
/// frame still parses — the message arrives looking fine and says something the peer never sent.
///
/// # Errors
/// [`TransportError::Decode`] if the frame is not valid UTF-8, or is not a message of this type.
/// Broken UTF-8 is a refusal, never a lossy accept.
pub fn decode_frame<T: DeserializeOwned>(frame: &[u8]) -> Result<T, TransportError> {
    let text = std::str::from_utf8(frame).map_err(|e| {
        TransportError::Decode(<serde_json::Error as serde::de::Error>::custom(format!(
            "a frame was not valid UTF-8 ({e})"
        )))
    })?;
    decode_line(text)
}

/// A [`Transport`] over any pair of byte streams.
///
/// Generic over the streams rather than tied to a socket, so the suite can drive it over in-memory
/// buffers. `R` is buffered because framing needs to read up to a delimiter.
pub struct NdjsonTransport<R: BufRead, W: Write, Out: Serialize, In: DeserializeOwned> {
    reader: R,
    writer: W,
    /// The frame being assembled, as bytes. Never a `String`: see [`decode_frame`] for why text
    /// conversion may only happen once a whole frame is in hand.
    frame: Vec<u8>,
    outbound: PhantomData<Out>,
    inbound: PhantomData<In>,
}

impl<R: BufRead, W: Write, Out: Serialize, In: DeserializeOwned> NdjsonTransport<R, W, Out, In> {
    /// Wrap a reader and a writer. Nothing is read or written until [`Transport::send`] or
    /// [`Transport::recv`] is called.
    pub fn new(reader: R, writer: W) -> Self {
        Self {
            reader,
            writer,
            frame: Vec::new(),
            outbound: PhantomData,
            inbound: PhantomData,
        }
    }

    /// Give back the streams, so a test can inspect exactly what was written.
    pub fn into_inner(self) -> (R, W) {
        (self.reader, self.writer)
    }
}

impl<R: BufRead, W: Write, Out: Serialize, In: DeserializeOwned> Transport
    for NdjsonTransport<R, W, Out, In>
{
    type Outbound = Out;
    type Inbound = In;

    fn send(&mut self, message: &Out) -> Result<(), TransportError> {
        let line = encode_line(message)?;
        self.writer.write_all(line.as_bytes())?;
        // Flushed per message: this protocol is live, and a frame held in a buffer until the next
        // one pushes it out is a meter that reads one frame late.
        self.writer.flush()?;
        Ok(())
    }

    fn recv(&mut self) -> Result<Option<In>, TransportError> {
        self.frame.clear();
        loop {
            let available = self.reader.fill_buf()?;
            if available.is_empty() {
                // End of stream. A partial frame here is a truncated one, so it decodes and fails
                // rather than returning a quiet `None`: discarding half a message silently is how a
                // client ends up rendering a world that was never sent.
                if self.frame.is_empty() {
                    return Ok(None);
                }
                return decode_frame(&self.frame).map(Some);
            }
            let (chunk, found) = match available.iter().position(|b| *b == DELIMITER) {
                Some(at) => (&available[..at], true),
                None => (available, false),
            };
            // The limit is in bytes, which is what a frame is made of and what a hostile peer
            // spends.
            if self.frame.len() + chunk.len() > MAX_LINE_BYTES {
                return Err(TransportError::FrameTooLarge {
                    limit: MAX_LINE_BYTES,
                });
            }
            // Bytes in, bytes out: a read boundary is an OS artifact and must not be able to change
            // a message's contents. Text conversion happens once, in `decode_frame`.
            self.frame.extend_from_slice(chunk);
            let consumed = chunk.len() + usize::from(found);
            self.reader.consume(consumed);
            if found {
                return decode_frame(&self.frame).map(Some);
            }
        }
    }
}
